api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/GenericCohortQueries.java

```diff
@@ -204,27 +205,123 @@ public CohortDefinition getPatientsBasedOnPatientStatesBeforeDate(int program, i
    * @return CohortDefinition
    */
   public CohortDefinition getDeceasedPatients() {
-    CompositionCohortDefinition cd = new CompositionCohortDefinition();
+    SqlCohortDefinition cd = new SqlCohortDefinition();
     cd.setName("Get deceased patients based on patient states and person object");
-    cd.addParameter(new Parameter("startDate", "Start Date", Date.class));
-    cd.addParameter(new Parameter("endDate", "End Date", Date.class));
+    cd.addParameter(new Parameter("onOrBefore", "End Date", Date.class));
     cd.addParameter(new Parameter("location", "Location", Location.class));
+    String sql =
+        "SELECT patient_id "
+            + "FROM   (SELECT patient_id, "
+            + "               Max(death_date) death_date "
+            + "        FROM   (SELECT p.patient_id, "
+            + "                       ps.start_date death_date "
+            + "                FROM   patient p "
+            + "                       JOIN patient_program pp "
+            + "                         ON p.patient_id = pp.patient_id "
+            + "                       JOIN patient_state ps "
+            + "                         ON pp.patient_program_id = ps.patient_program_id "
+            + "                WHERE  p.voided = 0 "
+            + "                       AND pp.voided = 0 "
+            + "                       AND pp.program_id = %d "
+            + "                       AND pp.location_id = :location "
+            + "                       AND ps.voided = 0 "
+            + "                       AND ps.state = %d "
+            + "                       AND ps.start_date <= :onOrBefore "
+            + "                       AND ps.end_date IS NULL "
+            + "                UNION "
+            + "                SELECT p.person_id, "
+            + "                       p.death_date "
+            + "                FROM   person p "
+            + "                WHERE  p.voided = 0"
```

Comment:
Should there be a condition here for `death_date <=:endDate` ?

Reply:
done